### PR TITLE
ATD for project metadata

### PR DESCRIPTION
**Pull request by aryx**
- [x] I ran `make setup && make` to update the generated code after editing a `.atd` file (TODO: have a CI check)

### CODE REVIEW

File: semgrep_output_v1.atd [hunk 1]
```diff
@@ -1006,49 +1006,49 @@ type project_metadata = {
      * TODO: see Github_metadata.ml
```
Comment (aryx) [inline semgrep_output_v1.atd:1006]: @zyannes would love some answers to the TODO comments I've put 10 lines above.

Comment (zyannes) [reply semgrep_output_v1.atd:1006]: This data is sent to the POST `/scans` endpoint when a scan is created and when getting the scan config.

Comment (zyannes) [reply semgrep_output_v1.atd:1006]: And a big +1 to splitting this into subtypes where `project_metadata` is a superset.

Comment (zyannes) [reply semgrep_output_v1.atd:1006]: FYI I don't think we are using the `project_metadata` type directly in the `/scans` endpoint today.

**Review by zyannes — APPROVED** [review]

File: semgrep_output_v1.proto [hunk 2]
```diff
@@ -340,27 +340,27 @@ message ParsingStats {
...
-  google.protobuf.Any commit_title = 23517072;
+  string repo_url = 245057343;
+  string branch = 387879587;
+  string ci_job_url = 514550082;
+  string commit = 200081263;
+  string commit_author_email = 458576327;
+  string commit_author_name = 534568416;
+  string commit_author_username = 247483738;
+  string commit_author_image_url = 314681517;
+  string commit_title = 23517072;
```
Comment (zyannes) [inline semgrep_output_v1.proto:351]: These should be `optional string`, right?

**Review by jbergler — COMMENTED** [review]
The ask I had for Matthew when he worked on similar changes recently was to summarize in the PR description the usage of all of these fields at the minumum supported version as for the reviewer to audit before approving.

@aryx do you mind going back to 1.17.0 and looking at the usages cli/src/ so I can review it?

File: semgrep_output_v1.proto (hunk 2 above)
Comment (aryx) [reply semgrep_output_v1.proto:351]: I don't know the semantic of protobuf, nor the jsonschema2protobuf script, but I guess yes. This is what is in the .jsonschema:
```
        "commit_author_email": { "type": [ "string", "null" ] },
        "commit_author_name": { "type": [ "string", "null" ] },
        "commit_author_username": { "type": [ "string", "null" ] },
        "commit_author_image_url": { "type": [ "string", "null" ] },
```

**Comment by aryx** [comment]
Here is the diff for meta.py between 1.17.0 and now:
```
diff -u -p -b -B -r -x .semantic.cache -x .depend -x CVS -x .hg -x .svn -x .git -x _darcs /home/pad/github/FRESH-semgrep/cli/src/semgrep/meta.py /tmp/meta2.py
--- /home/pad/github/FRESH-semgrep/cli/src/semgrep/meta.py	2023-09-26 08:12:33.936815327 +0200
+++ /tmp/meta2.py	2023-09-26 08:12:06.956813899 +0200
@@ -4,6 +4,7 @@ import subprocess
 import urllib.parse
 from dataclasses import dataclass
 from dataclasses import field
+from datetime import datetime
 from pathlib import Path
 from typing import Any
 from typing import Dict
@@ -165,6 +166,13 @@ class GitMeta:
         return git_check_output(["git", "show", "-s", "--format=%ct"])
 
     @property
+    def commit_timestamp(self) -> str:
+        """
+        Returns the commit timestamp as an iso-formatted datetime string.
+        """
+        return datetime.fromtimestamp(int(self.commit_datetime)).isoformat()
+
+    @property
     def is_full_scan(self) -> bool:
         return self.merge_base_ref is None
 
@@ -187,6 +195,7 @@ class GitMeta:
             "commit_author_username": None,
             "commit_author_image_url": None,
             "commit_title": commit_title,
+            "commit_timestamp": self.commit_timestamp,
             "on": self.event_name,
             "pull_request_author_username": None,
             "pull_request_author_image_url": None,
@@ -226,7 +235,14 @@ class GithubMeta(GitMeta):
 
     @property
     def repo_name(self) -> str:
-        repo_name = os.getenv("GITHUB_REPOSITORY")
+        repo_name = os.getenv("SEMGREP_REPO_NAME", os.getenv("GITHUB_REPOSITORY"))
+
+        # NOTE: no-super-first
+        # we don't call super().repo_url _first_ because that makes a call to the git remote
+        # if the env var is not set
+        # we want to prioritize the explicit env var, but the case where it isn't set,
+        # we should just do what we'd normally do
+        # in this particular case, we actually just don't call it at all
         if repo_name:
             return repo_name
         else:
@@ -235,8 +251,12 @@ class GithubMeta(GitMeta):
     @property
     def repo_url(self) -> Optional[str]:
         server_url = os.getenv("GITHUB_SERVER_URL", "https://github.com")
+        semgrep_repo_url = os.getenv("SEMGREP_REPO_URL")
 
-        if self.repo_name:
+        # see NOTE: no-super-first
+        if semgrep_repo_url:
+            return semgrep_repo_url
+        elif self.repo_name:
             return f"{server_url}/{self.repo_name}"
         return None
 
@@ -503,6 +523,10 @@ class GithubMeta(GitMeta):
 
     @property
     def ci_job_url(self) -> Optional[str]:
+        semgrep_job_url = super().ci_job_url
+        if semgrep_job_url:
+            return semgrep_job_url
+
         value = os.getenv("GITHUB_RUN_ID")
         if self.repo_url and value:
             return f"{self.repo_url}/actions/runs/{value}"
@@ -514,11 +538,19 @@ class GithubMeta(GitMeta):
 
     @property
     def pr_id(self) -> Optional[str]:
+        semgrep_pr_id = super().pr_id
+        if semgrep_pr_id:
+            return semgrep_pr_id
+
         pr_id = self.glom_event(T["pull_request"]["number"])
         return str(pr_id) if pr_id else None
 
     @property
     def pr_title(self) -> Optional[str]:
+        semgrep_pr_title = super().pr_title
+        if semgrep_pr_title:
+            return semgrep_pr_title
+
         pr_title = self.glom_event(T["pull_request"]["title"])
         return str(pr_title) if pr_title else None
 
@@ -545,7 +577,10 @@ class GithubMeta(GitMeta):
         """
         if self.event_name == "pull_request_target":
             return os.getenv("GITHUB_HEAD_REF")
-        return os.getenv("GITHUB_REF")
+        github_ref = os.getenv("SEMGREP_BRANCH", os.getenv("GITHUB_REF"))
+        if github_ref:
+            return github_ref
+        return super().branch
 
     def to_dict(self) -> Dict[str, Any]:
         return {
@@ -593,15 +628,30 @@ class GitlabMeta(GitMeta):
 
     @property
     def repo_name(self) -> str:
-        return os.getenv("CI_PROJECT_PATH", "[unknown]")
+        project_path = os.getenv("SEMGREP_REPO_NAME", os.getenv("CI_PROJECT_PATH"))
+
+        # see NOTE: no-super-first
+        if project_path:
+            return project_path
+        return super().repo_name
 
     @property
     def repo_url(self) -> Optional[str]:
-        return os.getenv("CI_PROJECT_URL")
+        project_url = os.getenv("SEMGREP_REPO_URL", os.getenv("CI_PROJECT_URL"))
+
+        # see NOTE: no-super-first
+        if project_url:
+            return project_url
+        return super().repo_url
 
     @property
     def commit_sha(self) -> Optional[str]:
-        return os.getenv("CI_COMMIT_SHA")
+        commit_sha = os.getenv("SEMGREP_COMMIT", os.getenv("CI_COMMIT_SHA"))
+
+        # see NOTE: no-super-first
+        if commit_sha:
+            return commit_sha
+        return super().commit_sha
 
     @property
     def commit_ref(self) -> Optional[str]:
@@ -620,7 +670,12 @@ class GitlabMeta(GitMeta):
 
     @property
     def ci_job_url(self) -> Optional[str]:
-        return os.getenv("CI_JOB_URL")
+        job_url = os.getenv("SEMGREP_JOB_URL", os.getenv("CI_JOB_URL"))
+
+        # see NOTE: no-super-first
+        if job_url:
+            return job_url
+        return super().ci_job_url
 
     @property
     def event_name(self) -> str:
@@ -631,7 +686,14 @@ class GitlabMeta(GitMeta):
 
     @property
     def pr_id(self) -> Optional[str]:
-        return os.getenv("CI_MERGE_REQUEST_IID")
+        pr_id = os.getenv("CI_MERGE_REQUEST_IID")
+        semgrep_pr_id = super().pr_id
+
+        if semgrep_pr_id:
+            return semgrep_pr_id
+        if pr_id:
+            return pr_id
+        return None
 
     @property
     def start_sha(self) -> Optional[str]:
@@ -639,7 +701,14 @@ class GitlabMeta(GitMeta):
 
     @property
     def pr_title(self) -> Optional[str]:
-        return os.getenv("CI_MERGE_REQUEST_TITLE")
+        pr_title = os.getenv("CI_MERGE_REQUEST_TITLE")
+        semgrep_pr_title = super().pr_title
+
+        if semgrep_pr_title:
+            return semgrep_pr_title
+        if pr_title:
+            return pr_title
+        return None
 
     def to_dict(self) -> Dict[str, Any]:
         return {

Diff finished.  Tue Sep 26 08:13:41 2023
```

**Comment by aryx** [comment]
So looks like the only relevant diff is commit_timestamp

**Comment by aryx** [comment]
I'll make it an optional field.

Good call.

**Comment by aryx** [comment]
was added in https://github.com/returntocorp/semgrep/pull/8542

**Pull request merged**